aptos-move/framework/supra-framework/sources/automation_registry.move

```diff
@@ -0,0 +1,356 @@
+    struct UpdateAutomationGasLimit has drop, store {
+        automation_gas_limit: u64
+    }
+
+    #[event]
+    /// Update duration upper limit event
+    struct UpdateDurationUpperLimit has drop, store {
+        duration_upper_limit: u64
+    }
+
+    #[event]
+    /// Remove automation task registry event
+    struct RemoveAutomationTask has drop, store {
+        id: u64
+    }
+
+    // todo : this function should call during initialzation, but since we already done genesis in that case who can access the function
+    public(friend) fun initialize(supra_framework: &signer) {
+        system_addresses::assert_supra_framework(supra_framework);
+
+        let (registry_fee_resource_signer, registry_fee_address_signer_cap) = account::create_resource_account(
+            supra_framework,
+            REGISTRY_RESOURCE_SEED
+        );
+
+        move_to(supra_framework, AutomationRegistry {
+            current_index: 0,
+            automation_gas_limit: DEFAULT_AUTOMATION_GAS_LIMIT,
+            duration_upper_limit: DEFAULT_DURATION_UPPER_LIMIT,
+            gas_committed_for_next_epoch: 0,
+            automation_unit_price: DEFAULT_AUTOMATION_UNIT_PRICE,
+            registry_fee_address: signer::address_of(&registry_fee_resource_signer),
+            registry_fee_address_signer_cap,
+            tasks: enumerable_map::new_map(),
+        })
+    }
+
+    public(friend) fun on_new_epoch(supra_framework: &signer) acquires AutomationRegistry {
+        system_addresses::assert_supra_framework(supra_framework);
+
+        let automation_registry = borrow_global_mut<AutomationRegistry>(@supra_framework);
+        let ids = enumerable_map::get_map_list(&automation_registry.tasks);
+
+        let current_time = timestamp::now_seconds();
+        let epoch_interval = block::get_epoch_interval_secs();
+        let expired_task_gas = 0;
+
+        // Perform clean up and updation of state
+        vector::for_each(ids, |id| {
+            let task = enumerable_map::get_value_mut(&mut automation_registry.tasks, id);
+
+            // Tasks that are active during this new epoch but will be already expired for the next epoch
+            if (task.expiry_time < (current_time + epoch_interval)) {
+                expired_task_gas = expired_task_gas + task.max_gas_amount;
+            };
+
+            if (task.expiry_time < current_time) {
+                enumerable_map::remove_value(&mut automation_registry.tasks, id);
+            } else if (!task.is_active && task.expiry_time > current_time) {
+                task.is_active = true;
+            }
+        });
+
+        // Adjust the gas committed for the next epoch by subtracting the gas amount of the expired task
+        automation_registry.gas_committed_for_next_epoch = automation_registry.gas_committed_for_next_epoch - expired_task_gas;
+    }
+
+    /// Withdraw accumulated automation task fees from the resource account
+    entry fun withdraw_automation_task_fees(
```

Comment:
maybe we can call this function refund?

Reply:
Actually I have replaced `withdraw_fee` function to `withdraw_automation_task_fees` which is only access by supra framework admin, it's not used for refund